# FastAPI

## Introducción

`FastAPI` está construido sobre otros frameworks:

 * `Uvicorn`: es una librería de Python que funciona de servidor, es decir, permite que cualquier computadora se convierta en un servidor
 * `Starlette`: es un framework de desarrollo web de bajo nivel, para desarrollar aplicaciones con este requieres un amplio conocimiento de Python, entonces FastAPI se encarga de añadirle funcionalidades por encima para que se pueda usar mas fácilmente
 * `Pydantic`: Es un framework que permite trabajar con datos similar a pandas, pero este te permite usar modelos los cuales aprovechara FastAPI para crear la API
 * `OpenAPI`: Es un conjunto de reglas que permite definir cómo describir, crear y visualizar APIs. Es un conjunto de reglas que permiten decir que una API está bien definida.

## Instalación

Para instalar `FastAPI` voy a hacerlo con `mamba`. En mi [post](https://maximofn.com/conda) anterior explico las ventajas de usar `conda` frente a `pip` y `mamba` para acelerar

Como hemos dicho `FastAPI` está construido sobre `uvicorn`, por lo que necesitamos instalarnos los dos

In [2]:
!mamba install -y -c conda-forge fastapi uvicorn


                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (1.3.1) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████


Looking for: ['fastapi', 'uvicorn']
...
Preparing transaction: done
Verifying transaction: done
Executing transac

## Hola Mundo

Para hacer todo nuestras APIs vamos a crear una carpeta llamada `FastAPI`

In [4]:
!mkdir FastAPI

Ahora creamos un hola mundo de `FastAPI`

In [5]:
%%writefile FastAPI/hello_world.py

from fastapi import FastAPI

app = FastAPI()

@app.get("/")
def home():
    return {"Hello": "World"}

Writing FastAPI/hello_world.py


Lo que hemos hecho es importar `FastAPI`, a cotinuación crear un objeto de la clase `FastAPI` que será nuestra aplicación y pur último la función que se ejecutará al acceder a la API. Más adelante explicaremos bien todo lo que hay en la función

Para poder lanzar la API necesitamos usar `uvicorn`, para ello hacemos `uvicorn FastAPI.hello_world:app --reload`. Ponemos `hello_world` porque hemos creado el archivo `hello_world.py`, `app` porque es cómo hemos llamado al objeto de la clase `FastAPI`, y `--reload` para que si hacemos cambios no tengamos qu relanzar la API

In [8]:
!uvicorn FastAPI.hello_world:app --reload

INFO:     Will watch for changes in these directories: ['/home/wallabot/Documentos/web/portafolio/posts']
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [55677] using StatReload
INFO:     Started server process [55679]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     127.0.0.1:51580 - "GET / HTTP/1.1" 200 OK
INFO:     127.0.0.1:44282 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:44282 - "GET /openapi.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:39108 - "GET /redoc HTTP/1.1" 200 OK
INFO:     127.0.0.1:39108 - "GET /openapi.json HTTP/1.1" 200 OK


Como vemos, nos da un enlace [http://127.0.0.1:8000](http://127.0.0.1:8000), si accedemos a él a través del navegador, podemos leer `{"Hello":"World"}`

## Documentación interactiva

Como hemos dicho `FastAPI` está construido también sobre `OpenAPI` que es un conjunto de reglas que definen cómo crear APIs

`OpenAPI` necesita de un software, el cual es `Swagger`, que es un conjunto de softwares que permiten trabajar con APIs. `FastAPI` funciona sobre un programa de `Swagger` el cual es `Swagger UI`, que permite mostrar la API documentada.

Acceder a la documentación interactiva con `Swagger UI`:

`<IP>/docs`


Acceder a la documentación interactiva con `Redoc`:

`<IP>/redoc`

Si volvemos a lanzar la API de hello world

In [9]:
!uvicorn FastAPI.hello_world:app --reload

INFO:     Will watch for changes in these directories: ['/home/wallabot/Documentos/web/portafolio/posts']
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [56013] using StatReload
INFO:     Started server process [56015]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     127.0.0.1:57266 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:57266 - "GET /openapi.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:42336 - "GET /openapi.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:42338 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:42338 - "GET /openapi.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:57620 - "GET / HTTP/1.1" 200 OK


Y accedemos a [http://127.0.0.1:8000/docs](http://127.0.0.1:8000/docs) o [http://127.0.0.1:8000/redoc](http://127.0.0.1:8000/redoc) podremos ver la documentación que nos ha generado `FastAPI` automáticamente

hello_world/docs:

![hello_world/docs](https://maximofn.com/wp-content/uploads/2023/05/hello_world_docs.png)

hello_world/redoc:

![hello_world/redoc](https://maximofn.com/wp-content/uploads/2023/05/hello_world_redoc.png)

## Path operation

Path operation define las operaciones que podemos hacer sobre un path determinado. Si recuerdas en el ejemplo de hola mundo escribimos la siguiente función

``` python
@app.get("/")
def home():
    return {"Hello": "World"}
```

El decorador es la línea que comienza con `@` y que define que cuando hagamos mediante `HTTP` la operación de `GET` sobre el path `/` se va a ejecutar la función `home()`

Lo que es el path está claro, dentro de la API podemos definir carpetas y paths, pero las operaciones son las posibles operaciones `HTTP` que se pueden realizar y que son

 * `Get`: Traer la documentación
 * `Post`: Enviar documentación
 * `Put`: Actualizar información de la documentación que ya está en el servidor
 * `Delete`: Borrar información
 * `Options`: Devuelve un header adicional llamado allow que contiene los metodos http que pueden utilizarse en ese endpoint o path.
 * `Head`: Devuelve info sobre el documento, no el documento en si.
 * `Patch`: Hacer modificaciones parciales al documento a diferencia de put que permite cambiar el documento entero.
 * `Trace`: Nos permite observar que esta pasando en la peticion y nos devuelve nuestro input con propositos de debugging.

Volviendo al ejemplo del hola mundo

``` python
@app.get("/")
def home():
    return {"Hello": "World"}
```

La primera linea corresponde a la `path operation decorator` y el resto al `path operation function`

## Path parameters

Supongamos que estamos haciendo una API, para poder obtener los datos de un dataset para entrenar una red. Por ejemplo, para un problema de segmentación, podemos tener una carpeta por cada par de imagen y másara

```
dataset
 |--> Imagen1
       |-->Image
       |-->Mask
 |--> Imagen2
       |-->Image
       |-->Mask
 |--> Imagen3
       |-->Image
       |-->Mask
 ...
 |--> ImagenN
       |-->Image
       |-->Mask
```

Por lo que podemos crear los paths `/dataset/Imagen1`, `/dataset/Imagen2`, `/dataset/Imagen3`, ... . Pero si el dataset es variable? Si nunca vamos a saber cuántas imágenes tiene el dataset?

Por ello necesitamos poder crear el path mediante parámetros. Así, externamente podemos saber la longitud del dataset y crear el path que queramos

Para crear los path mediante parámetros se hace de la siguiente manera `/dataset/{Image}`, donde `{Image}` es el parámetro

Si usamos `path parameter` el paso del parámetro es obligatorio, por lo que si queremos usar algo parecido, pero que sea opcional tenemos que usar `query parameter`

## Query parameter

Supongamos ahora que estamos haciendo una API con una base de datos de usuarios, en la que cada usuario puede tener o no `username`, `age`, `email`. Los `query parameter`s se suelen utilizar más con la operación de `PUT`, es decir, mediante el path pasamos información del usuraio.

Puede que de muchos usuarios sepamos todos los datos, pero ¿qué pasa si un usuario no tiene email? Si hiciésemos uso de `path parameter`s no podríamos añadir ese usuario, o lo tedríamos que hacer con un email que previamente definamos como que el usuario no tiene email. Sin embargo, si usamos `query parameter`s no es obligatorio pasar todos los parámetros

La sintaxis de los `query parameter`s es `/path?key1=value1&key2=value2&key3=value3&...&keyN=valueN`

Por ejemplo, para la API con la base de datos de usuarios una `query parameter` podría ser `/path?username=Pedro&email=pedro@gmail.com&age=30` y otra solo `/path?username=Luis`

Podemos mezclar los dos (`path parameter`s y `query parameter`s) y mejorar los paths anteriores con `/path/{pedro_id}/details?username=Pedro&email=pedro@gmail.com&age=30` y otra solo `/path/{luis_id}/details?username=Luis`

## Request Body y Response Body

En una comunicación `HTTP` cliente/servidor, la petición que le hace el cliente al servidor se denomina `request` y la respuesta del servidor se denomina `response`. Además dentro del protocolo `HTTP` hay `header`s y el `body`. 

Por tanto, una `request body` es solo el `body` de una `request` `HTTP`, mientras que una `response body` es solo el `body` de una `response` `HTTP`

Esta `request body` y `response body` se hace entre el cliente y servidor mediante `JSON`s

Editamos nuestro hola mundo para ver cómo se programaría un `request body` y un `response body`

In [1]:
%%writefile FastAPI/hello_world.py

from fastapi import FastAPI

app = FastAPI()

@app.get("/")
def home():
    return {"Hello": "World"}

# Request and Response body
@app.post("/user/new")
def new_user():
    pass

Overwriting FastAPI/hello_world.py


Como vemos solo hemos creado una operación `POST` y su función asociada, de momento lo dejamos sin definir y vemos más conceptos

## Models

Un modelo es una descripción en código de un elemento de la vida real. Para poder crear nuestros modelos vamos a hacer uso de una librería llamada `pydantic`. Con esto vamos a crear un modelo de la clase `user`

In [2]:
%%writefile FastAPI/hello_world.py

from typing import Optional
from pydantic import BaseModel
from fastapi import FastAPI
from fastapi import Body

app = FastAPI()

class User(BaseModel):
    name: str
    age: int
    email: Optional[str] = None

@app.get("/")
def home():
    return {"Hello": "World"}

# Request and Response body
@app.post("/user/new")
def new_user(user: User = Body(...)):
    return user

Overwriting FastAPI/hello_world.py


Analizamos el código

Con la línea

``` python
from typing import Optional
```

Estamos añadiendo el módulo `Optional` de la librería `typing` que nos va a permitir añadir parámetros opcionales

Con la línea 

``` python
from pydantic import BaseModel
```

Estamos importando el módulo `BaseModel` de la librería `pydantic` que nos va a permitir crear clases a partir de ese modelo base

Con la línea 

``` python
from fastapi import Body
```

Importamos el módulo `Body` que nos permite usar objetos de tipo `Body`

Con el bloque

``` python
class User(BaseModel):
    name: str
    age: int
    email: Optional[str] = None
```

Definimos la clase `User` que hereda de `BaseModel` y que tiene los parámetros `name` y `age` obligatorios y el parámetro `email` que es opcional, y al ser opcional tenemos que darle un valor por defecto, que en este caso es `None`

Con el bloque

``` python
# Request and Response body
@app.post("/user/new")
def new_user(user: User = Body(...)):
    return user
```

definimos la función para la operación `POST` que añade un nuevo usuario. Los `...` indican que usuario es un parámetro obligatorio que hay que pasarle a la función

Levantamos la API

In [3]:
!uvicorn FastAPI.hello_world:app --reload

INFO:     Will watch for changes in these directories: ['/home/wallabot/Documentos/web/portafolio/posts']
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [21454] using StatReload
INFO:     Started server process [21456]
INFO:     Waiting for application startup.
INFO:     Application startup complete.


Si ahora hacemos un `POST` a esta aplicación con nuevo usuario

In [20]:
!curl -X 'POST' \
  'http://127.0.0.1:8000/user/new' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{"name": "Mariano","age": 30,"email": "mariano@email.com"}'

{"name":"Mariano","age":30,"email":"mariano@email.com"}

Qué pasa si en el método `POST` quiero mandar dos modelos

In [28]:
%%writefile FastAPI/hello_world.py

from typing import Optional
from pydantic import BaseModel
from fastapi import FastAPI
from fastapi import Body

app = FastAPI()

class User(BaseModel):
    name: str
    age: int
    email: Optional[str] = None

class Location(BaseModel):
    city: str
    state: str
    country: str

@app.get("/")
def home():
    return {"Hello": "World"}

# Request and Response body
@app.post("/user/new")
def new_user(
    user: User = Body(...), 
    location: Location = Body(...)
):
    return {"user": user, "location": location}

Overwriting FastAPI/hello_world.py


Como vemos ahora hay que mandarle a la API el usuario y la localización, por lo que en el `return` ha habido que juntar los dos diccionarios

Volvemos a levantar la API y a probar

In [22]:
!uvicorn FastAPI.hello_world:app --reload

INFO:     Will watch for changes in these directories: ['/home/wallabot/Documentos/web/portafolio/posts']
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [23730] using StatReload
INFO:     Started server process [23732]
INFO:     Waiting for application startup.
INFO:     Application startup complete.


In [29]:
!curl -X 'POST' \
  'http://127.0.0.1:8000/user/new' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{"user": {"name": "Mariano", "age": 30, "email": "mariano@email.com"}, "location": {"city": "Sevilla", "state": "Andalucía", "country": "España"} }'

{"user":{"name":"Mariano","age":30,"email":"mariano@email.com"},"location":{"city":"Sevilla","state":"Andalucía","country":"España"}}

Aunque otra forma de juntar los diccionarios es

In [33]:
%%writefile FastAPI/hello_world.py

from typing import Optional
from pydantic import BaseModel
from fastapi import FastAPI
from fastapi import Body

app = FastAPI()

class User(BaseModel):
    name: str
    age: int
    email: Optional[str] = None

class Location(BaseModel):
    city: str
    state: str
    country: str

@app.get("/")
def home():
    return {"Hello": "World"}

# Request and Response body
@app.post("/user/new")
def new_user(
    user: User = Body(...), 
    location: Location = Body(...)
):
    resutl = user.dict()
    resutl.update(location.dict())
    return resutl

Overwriting FastAPI/hello_world.py


In [27]:
!uvicorn FastAPI.hello_world:app --reload

INFO:     Will watch for changes in these directories: ['/home/wallabot/Documentos/web/portafolio/posts']
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [24346] using StatReload
INFO:     Started server process [24348]
INFO:     Waiting for application startup.
INFO:     Application startup complete.


In [34]:
!curl -X 'POST' \
  'http://127.0.0.1:8000/user/new' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{"user": {"name": "Mariano", "age": 30, "email": "mariano@email.com"}, "location": {"city": "Sevilla", "state": "Andalucía", "country": "España"} }'

{"name":"Mariano","age":30,"email":"mariano@email.com","city":"Sevilla","state":"Andalucía","country":"España"}

La diferencia con el ejemplo anterior es que se han juntado los dos diccionarios en uno solo

## Validaciones de los query parameters

Supongamos que a la hora de crear un usuario queremos establecer que el nombre tenga como máximo 50 caracteres y como mínimo 1 caracter, así que esto se puede programar

In [1]:
%%writefile FastAPI/hello_world.py

from typing import Optional
from pydantic import BaseModel
from fastapi import FastAPI
from fastapi import Body
from fastapi import Query

app = FastAPI()

class User(BaseModel):
    name: str
    age: int
    email: Optional[str] = None

class Location(BaseModel):
    city: str
    state: str
    country: str

@app.get("/")
def home():
    return {"Hello": "World"}

# Request and Response body
@app.post("/user/new")
def new_user(
    user: User = Body(...), 
    location: Location = Body(...)
):
    resutl = user.dict()
    resutl.update(location.dict())
    return resutl

# Validaciones de los query parameters
@app.get("/user/detail")
def user_show_detail(
    name: Optional[str] = Query(None, min_length=1, max_length=50),
    age: Optional[int] = Query(None, gt=0, le=100)
):
    return {"name": name, "age": age}

Overwriting FastAPI/hello_world.py


Hemos necesitado importar al inicio `Query` desde `fastapi`. A continuación hemos creado la función `user_show_detail` que espera dos `query parameter`s. EL primero es `name` que como mínimo tiene que tener 1 caracter y como máximo 50, y `age` que como mínimo tiene que valer 0 y como máximo 100

Lanzamos la API

In [2]:
!uvicorn FastAPI.hello_world:app --reload

INFO:     Will watch for changes in these directories: ['/home/wallabot/Documentos/web/portafolio/posts']
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [23814] using StatReload
INFO:     Started server process [23816]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     127.0.0.1:41110 - "GET / HTTP/1.1" 200 OK
INFO:     127.0.0.1:41110 - "GET /favicon.ico HTTP/1.1" 404 Not Found
INFO:     127.0.0.1:59194 - "GET / HTTP/1.1" 200 OK
INFO:     127.0.0.1:46092 - "GET /user/detail HTTP/1.1" 200 OK
INFO:     127.0.0.1:42002 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:42002 - "GET /openapi.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:44462 - "GET /user/detail?name=mariano&age=30 HTTP/1.1" 200 OK


Si hacemos un `GET` a `http://127.0.0.1:8000/user/detail`

In [1]:
!curl -X 'GET' \
  'http://127.0.0.1:8000/user/detail'

{"name":null,"age":null}

No obtenemos nada porque no hemos metido ni nombre ni edad, vamos a meter un nombre y una edad

In [2]:
!curl -X 'GET' \
  'http://127.0.0.1:8000/user/detail?name=Mariano&age=30'

{"name":"Mariano","age":30}

Ahora vemos que si nos devuelve un nombre y una edad

Probamos ahora a meter un nombre con más de 50 caracteres

In [4]:
!curl -X 'GET' \
  'http://127.0.0.1:8000/user/detail?name=012345678911234567892123456789312345678941234567895123456789&age=30'

{"detail":[{"loc":["query","name"],"msg":"ensure this value has at most 50 characters","type":"value_error.any_str.max_length","ctx":{"limit_value":50}}]}

Como vemos obtenemos un error que nos indica que no podemos meter más de 50 caaracteres

Lo mismo ocurre si metemos una edad de más de 100

In [5]:
!curl -X 'GET' \
  'http://127.0.0.1:8000/user/detail?name=Mariano&age=101'

{"detail":[{"loc":["query","age"],"msg":"ensure this value is less than or equal to 100","type":"value_error.number.not_le","ctx":{"limit_value":100}}]}

Nos dice que la edad límite es 100

Vamos a ver qué más parametros se puede validar además de la longitud y el valor máximo y mínimo:

 * Para tipos de datos str:
   * max_length : Para especificar el tamaño máximo de la cadena.
   * min_length : Para especificar el tamaño minimo de la cadena.
   * regex : Para especificar expresiones regulares.
 * Para tipos de datos int:
   * ge : (greater or equal than ≥) Para especificar que el valor debe ser mayor o igual.
   * le : (less or equal than ≤) Para especificar que el valor debe ser menor o igual.
   * gt : (greater than >) Para especificar que el valor debe ser mayor.
   * lt : (less than <) Para especificar que el valor debe ser menor.

Para darle un título y una descripción en la documentación están
 * title: Para crear un título
 * description: Para añadir una descripción

## Validación de path parameters

Al igual que los `query parameter`s se puede validar los `path parameter`s

In [6]:
%%writefile FastAPI/hello_world.py

from typing import Optional
from pydantic import BaseModel
from fastapi import FastAPI
from fastapi import Body
from fastapi import Query
from fastapi import Path

app = FastAPI()

class User(BaseModel):
    name: str
    age: int
    email: Optional[str] = None

class Location(BaseModel):
    city: str
    state: str
    country: str

@app.get("/")
def home():
    return {"Hello": "World"}

# Request and Response body
@app.post("/user/new")
def new_user(
    user: User = Body(...), 
    location: Location = Body(...)
):
    resutl = user.dict()
    resutl.update(location.dict())
    return resutl

# Validaciones de los query parameters
@app.get("/user/detail")
def user_show_detail(
    name: Optional[str] = Query(
        None, 
        min_length=1, 
        max_length=50,
        title="User name",
        description="This is the user name. It's between 1 and 50 characters"
        ),
    age: Optional[int] = Query(
        None, 
        gt=0, 
        le=100,
        title="User age",
        description="This is the user age. It's between 1 and 100"
        )
):
    return {"name": name, "age": age}

# Validaciones de los path parameters
@app.get("/user/detail/{user_id}")
def user_show_detail(
    user_id: int = Path(
        ..., 
        gt=0, 
        le=100,
        title="User ID",
        description="This is the user ID. It's between 1 and 100"
        )
):
    return {"user_id": user_id}

Overwriting FastAPI/hello_world.py


Importamos `Path` desde `fastapi`.

Por otro lado creamos la función `user_show_detail` que espera un `path parameter` en el endpoint `/user/detail/{user_id}`. La función espera un `ID` por lo que espera un entero. Al ser un `path parameter` ponemos `...` porque tiene que ser un parámetro obligatorio y además decimos que tiene que ser mayor que 0 y menor que 100

Aprovechamos y ponemos `title` y `description` en las dos funciones `user_show_detail` para que aparezcan en la documentación

Levantamos la API

In [7]:
!uvicorn FastAPI.hello_world:app --reload

INFO:     Will watch for changes in these directories: ['/home/wallabot/Documentos/web/portafolio/posts']
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [33520] using StatReload
INFO:     Started server process [33522]
INFO:     Waiting for application startup.
INFO:     Application startup complete.


Probamos ahora a hacer un `GET` a `http://127.0.0.1:8000/user/detail/5`, es decir, le pasamos `5` como `ID`

In [10]:
!curl -X 'GET' \
  'http://127.0.0.1:8000/user/detail/5'

{"user_id":5}

Nos devuelve la `ID`

Probamos ahora a pasarle una `ID` de `101`

In [11]:
!curl -X 'GET' \
  'http://127.0.0.1:8000/user/detail/101'

{"detail":[{"loc":["path","user_id"],"msg":"ensure this value is less than or equal to 100","type":"value_error.number.not_le","ctx":{"limit_value":100}}]}

Nos da un error y nos dice que como máximo puede ser 100

## Validación de los Models

Vamos a validar los `Model`s

In [6]:
%%writefile FastAPI/hello_world.py

from typing import Optional
from pydantic import BaseModel
from pydantic import Field
from pydantic import validator
from enum import Enum
from fastapi import FastAPI
from fastapi import Body
from fastapi import Query
from fastapi import Path

app = FastAPI()

class ComunidadAutonoma(str, Enum):
    Andalucía = "Andalucía"
    Aragón = "Aragón"
    Asturias = "Asturias"
    Baleares = "Baleares"
    Canarias = "Canarias"
    Cantabria = "Cantabria"
    CastillaLaMancha = "Castilla-La Mancha"
    CastillaLeon = "Castilla y León"
    Cataluña = "Cataluña"
    Ceuta = "Ceuta"
    Extremadura = "Extremadura"
    Galicia = "Galicia"
    Madrid = "Madrid"
    Melilla = "Melilla"
    Murcia = "Murcia"
    Navarra = "Navarra"
    PaisVasco = "País Vasco"
    LaRioja = "La Rioja"
    Valencia = "Valencia"

class User(BaseModel):
    name: str = Field(..., min_length=1, max_length=50) # Required
    age: int = Field(..., gt=0, le=100) # Required
    email: Optional[str] = Field(None) # Optional

class Location(BaseModel):
    city: str = Field(..., min_length=1, max_length=50) # Required
    state: ComunidadAutonoma = Field(..., min_length=1, max_length=50) # Required
    country: str = Field(..., min_length=1, max_length=50) # Required

    @validator("state", pre=True)
    def validate_state(cls, state):
        if state not in ComunidadAutonoma._member_names_:
            raise ValueError(f"Invalid state: {state}")
        return state

@app.get("/")
def home():
    return {"Hello": "World"}

# Request and Response body
@app.post("/user/new")
def new_user(
    user: User = Body(...), 
    location: Location = Body(...)
):
    resutl = user.dict()
    resutl.update(location.dict())
    return resutl

# Validaciones de los query parameters
@app.get("/user/detail")
def user_show_detail(
    name: Optional[str] = Query(
        None, 
        min_length=1, 
        max_length=50,
        title="User name",
        description="This is the user name. It's between 1 and 50 characters"
        ),
    age: Optional[int] = Query(
        None, 
        gt=0, 
        le=100,
        title="User age",
        description="This is the user age. It's between 1 and 100"
        )
):
    return {"name": name, "age": age}

# Validaciones de los path parameters
@app.get("/user/detail/{user_id}")
def user_show_detail(
    user_id: int = Path(
        ..., 
        gt=0, 
        le=100,
        title="User ID",
        description="This is the user ID. It's between 1 and 100"
        )
):
    return {"user_id": user_id}

Overwriting FastAPI/hello_world.py


Importando la clase `Field` de `pydantic` obtenemos la herramienta para validar cada campo de un `Model`, de manera que en cada campo de los `Model`s se indica que es de tipo `Field` y se le indican las características

Además importando la clase `Enum` de `enum` podemos crear una clase con todas las posibles comunidades autónomas que puede tener el campo `state` de `Location`. Importando la clase `validator` de `pydantic` podemos validar que la comunidad autónoma esté dentro de uno de los posibles valores de la clase `ComunidadAutonoma` y esto es gracias al método `validate_state` de la clase `Location`

Vamos a levantar la API

In [2]:
!uvicorn FastAPI.hello_world:app --reload

INFO:     Will watch for changes in these directories: ['/home/wallabot/Documentos/web/portafolio/posts']
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [42849] using StatReload
INFO:     Started server process [42851]
INFO:     Waiting for application startup.
INFO:     Application startup complete.


Mando datos correctos

In [3]:
!curl -X 'POST' \
  'http://127.0.0.1:8000/user/new' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{"user": {"name": "Mariano", "age": 30, "email": "mariano@email.com"}, "location": {"city": "Sevilla", "state": "Andalucía", "country": "España"} }'

{"name":"Mariano","age":30,"email":"mariano@email.com","city":"Sevilla","state":"Andalucía","country":"España"}

Pruebo ahora a mandar un dato incorrecto, como la edad de 101 años

In [7]:
!curl -X 'POST' \
  'http://127.0.0.1:8000/user/new' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{"user": {"name": "Mariano", "age": 101, "email": "mariano@email.com"}, "location": {"city": "Sevilla", "state": "Andalucía", "country": "España"} }'

{"detail":[{"loc":["body","user","age"],"msg":"ensure this value is less than or equal to 100","type":"value_error.number.not_le","ctx":{"limit_value":100}}]}

Como se ve nos dice que la edad tiene que ser menor o igual a 100 años

Probamos ahora a meter una comunidad autónoma que no esté en la clase `ComunidadAutonoma`

In [10]:
!curl -X 'POST' \
  'http://127.0.0.1:8000/user/new' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{"user": {"name": "Mariano", "age": 30, "email": "mariano@email.com"}, "location": {"city": "Sevilla", "state": "AnWalucía", "country": "España"} }'

{"detail":[{"loc":["body","location","state"],"msg":"Invalid state: AnWalucía","type":"value_error"}]}

## Valores por defecto en la documentación

Con lo que hemos hecho hasta ahora, si nos vamos a la documentación vemos esto en la operación `POST`

![post default](https://maximofn.com/wp-content/uploads/2023/05/hello_world_docs_post_default.png)

Es decir, no tenemos valores por defecto, por lo que si queremos probar la operación tenemos que meter a mano uno a uno. Esto cuando se está desarrollando es un lastre, porque hay que mater todos a mano, pruebas, si no funciona cambias el código, recargas la página y tienes que volver a meter los valores a mano

Para solucionar esto, podemos crear una clase dentro de las clases `User` y `Location` llamada `Config` con los valores por defecto, veamos el código

In [7]:
%%writefile FastAPI/hello_world.py

from typing import Optional
from pydantic import BaseModel
from pydantic import Field
from pydantic import validator
from enum import Enum
from fastapi import FastAPI
from fastapi import Body
from fastapi import Query
from fastapi import Path

app = FastAPI()

class ComunidadAutonoma(str, Enum):
    Andalucía = "Andalucía"
    Aragón = "Aragón"
    Asturias = "Asturias"
    Baleares = "Baleares"
    Canarias = "Canarias"
    Cantabria = "Cantabria"
    CastillaLaMancha = "Castilla-La Mancha"
    CastillaLeon = "Castilla y León"
    Cataluña = "Cataluña"
    Ceuta = "Ceuta"
    Extremadura = "Extremadura"
    Galicia = "Galicia"
    Madrid = "Madrid"
    Melilla = "Melilla"
    Murcia = "Murcia"
    Navarra = "Navarra"
    PaisVasco = "País Vasco"
    LaRioja = "La Rioja"
    Valencia = "Valencia"

class User(BaseModel):
    name: str = Field(..., min_length=1, max_length=50) # Required
    age: int = Field(..., gt=0, le=100) # Required
    email: Optional[str] = Field(None) # Optional

    class Config:
        schema_extra = {
            "example": {
                "name": "Mariano",
                "age": 30,
                "email": "mariano@email.com"
            },
        }

class Location(BaseModel):
    city: str = Field(..., min_length=1, max_length=50) # Required
    state: ComunidadAutonoma = Field(..., min_length=1, max_length=50) # Required
    country: str = Field(..., min_length=1, max_length=50) # Required

    @validator("state", pre=True)
    def validate_state(cls, state):
        if state not in ComunidadAutonoma._member_names_:
            raise ValueError(f"Invalid state: {state}")
        return state
    
    class Config:
        schema_extra = {
            "example": {
                "city": "Sevilla",
                "state": "Andalucía",
                "country": "España"
            },
        }

@app.get("/")
def home():
    return {"Hello": "World"}

# Request and Response body
@app.post("/user/new")
def new_user(
    user: User = Body(...), 
    location: Location = Body(...)
):
    resutl = user.dict()
    resutl.update(location.dict())
    return resutl

# Validaciones de los query parameters
@app.get("/user/detail")
def user_show_detail(
    name: Optional[str] = Query(
        None, 
        min_length=1, 
        max_length=50,
        title="User name",
        description="This is the user name. It's between 1 and 50 characters"
        ),
    age: Optional[int] = Query(
        None, 
        gt=0, 
        le=100,
        title="User age",
        description="This is the user age. It's between 1 and 100"
        )
):
    return {"name": name, "age": age}

# Validaciones de los path parameters
@app.get("/user/detail/{user_id}")
def user_show_detail(
    user_id: int = Path(
        ..., 
        gt=0, 
        le=100,
        title="User ID",
        description="This is the user ID. It's between 1 and 100"
        )
):
    return {"user_id": user_id}

Overwriting FastAPI/hello_world.py


Una vez hecho esto, la documentación se ve así 

![docs with values](https://maximofn.com/wp-content/uploads/2023/05/hello_world_docs_POST_values.png)

Ahora tenemos todos los campos rellenos, por lo que si queremos probar no tenemos que meterlos nosotros y nos ahorra mucho tiempo a la hora de depurar

Con las clases `User` y `Location` podemos establecer valores por defecto para la documentación, lo cual nos cubre el método `POST`, pero para las dos funciones `GET` `user_show_detail` no nos vale ya que no usan las clases `User` y `Location`. Así que en ese caso, podemos establecer el valor por defecto para la documentación añadiendo el parámetro `example` en `name`, `age` y `user_id`

In [2]:
%%writefile FastAPI/hello_world.py

from typing import Optional
from pydantic import BaseModel
from pydantic import Field
from pydantic import validator
from enum import Enum
from fastapi import FastAPI
from fastapi import Body
from fastapi import Query
from fastapi import Path

app = FastAPI()

class ComunidadAutonoma(str, Enum):
    Andalucía = "Andalucía"
    Aragón = "Aragón"
    Asturias = "Asturias"
    Baleares = "Baleares"
    Canarias = "Canarias"
    Cantabria = "Cantabria"
    CastillaLaMancha = "Castilla-La Mancha"
    CastillaLeon = "Castilla y León"
    Cataluña = "Cataluña"
    Ceuta = "Ceuta"
    Extremadura = "Extremadura"
    Galicia = "Galicia"
    Madrid = "Madrid"
    Melilla = "Melilla"
    Murcia = "Murcia"
    Navarra = "Navarra"
    PaisVasco = "País Vasco"
    LaRioja = "La Rioja"
    Valencia = "Valencia"

class User(BaseModel):
    name: str = Field(..., min_length=1, max_length=50) # Required
    age: int = Field(..., gt=0, le=100) # Required
    email: Optional[str] = Field(None) # Optional

    class Config:
        schema_extra = {
            "example": {
                "name": "Mariano",
                "age": 30,
                "email": "mariano@email.com"
            },
        }

class Location(BaseModel):
    city: str = Field(..., min_length=1, max_length=50) # Required
    state: ComunidadAutonoma = Field(..., min_length=1, max_length=50) # Required
    country: str = Field(..., min_length=1, max_length=50) # Required

    @validator("state", pre=True)
    def validate_state(cls, state):
        if state not in ComunidadAutonoma._member_names_:
            raise ValueError(f"Invalid state: {state}")
        return state
    
    class Config:
        schema_extra = {
            "example": {
                "city": "Sevilla",
                "state": "Andalucía",
                "country": "España"
            },
        }

@app.get("/")
def home():
    return {"Hello": "World"}

# Request and Response body
@app.post("/user/new")
def new_user(
    user: User = Body(...), 
    location: Location = Body(...)
):
    resutl = user.dict()
    resutl.update(location.dict())
    return resutl

# Validaciones de los query parameters
@app.get("/user/detail")
def user_show_detail(
    name: Optional[str] = Query(
        None, 
        min_length=1, 
        max_length=50,
        title="User name",
        description="This is the user name. It's between 1 and 50 characters",
        example="Mariano"
        ),
    age: Optional[int] = Query(
        None, 
        gt=0, 
        le=100,
        title="User age",
        description="This is the user age. It's between 1 and 100",
        example=30
        )
):
    return {"name": name, "age": age}

# Validaciones de los path parameters
@app.get("/user/detail/{user_id}")
def user_show_detail(
    user_id: int = Path(
        ..., 
        gt=0, 
        le=100,
        title="User ID",
        description="This is the user ID. It's between 1 and 100",
        example=1
        )
):
    return {"user_id": user_id}

Overwriting FastAPI/hello_world.py


## Response Model

Al código anterior vamos a añadirle a la clase `User` el atributo `password`

In [4]:
%%writefile FastAPI/hello_world.py

from typing import Optional
from pydantic import BaseModel
from pydantic import Field
from pydantic import validator
from enum import Enum
from fastapi import FastAPI
from fastapi import Body
from fastapi import Query
from fastapi import Path

app = FastAPI()

class ComunidadAutonoma(str, Enum):
    Andalucía = "Andalucía"
    Aragón = "Aragón"
    Asturias = "Asturias"
    Baleares = "Baleares"
    Canarias = "Canarias"
    Cantabria = "Cantabria"
    CastillaLaMancha = "Castilla-La Mancha"
    CastillaLeon = "Castilla y León"
    Cataluña = "Cataluña"
    Ceuta = "Ceuta"
    Extremadura = "Extremadura"
    Galicia = "Galicia"
    Madrid = "Madrid"
    Melilla = "Melilla"
    Murcia = "Murcia"
    Navarra = "Navarra"
    PaisVasco = "País Vasco"
    LaRioja = "La Rioja"
    Valencia = "Valencia"

class User(BaseModel):
    name: str = Field(..., min_length=1, max_length=50) # Required
    age: int = Field(..., gt=0, le=100) # Required
    email: Optional[str] = Field(None) # Optional
    password: str = Field(..., min_length=8, max_length=50) # Required

    class Config:
        schema_extra = {
            "example": {
                "name": "Mariano",
                "age": 30,
                "email": "mariano@email.com",
                "password": "12345678"
            },
        }

class Location(BaseModel):
    city: str = Field(..., min_length=1, max_length=50) # Required
    state: ComunidadAutonoma = Field(..., min_length=1, max_length=50) # Required
    country: str = Field(..., min_length=1, max_length=50) # Required

    @validator("state", pre=True)
    def validate_state(cls, state):
        if state not in ComunidadAutonoma._member_names_:
            raise ValueError(f"Invalid state: {state}")
        return state
    
    class Config:
        schema_extra = {
            "example": {
                "city": "Sevilla",
                "state": "Andalucía",
                "country": "España"
            },
        }

@app.get("/")
def home():
    return {"Hello": "World"}

# Request and Response body
@app.post("/user/new")
def new_user(
    user: User = Body(...), 
    location: Location = Body(...)
):
    resutl = user.dict()
    resutl.update(location.dict())
    return resutl

# Validaciones de los query parameters
@app.get("/user/detail")
def user_show_detail(
    name: Optional[str] = Query(
        None, 
        min_length=1, 
        max_length=50,
        title="User name",
        description="This is the user name. It's between 1 and 50 characters",
        example="Mariano"
        ),
    age: Optional[int] = Query(
        None, 
        gt=0, 
        le=100,
        title="User age",
        description="This is the user age. It's between 1 and 100",
        example=30
        )
):
    return {"name": name, "age": age}

# Validaciones de los path parameters
@app.get("/user/detail/{user_id}")
def user_show_detail(
    user_id: int = Path(
        ..., 
        gt=0, 
        le=100,
        title="User ID",
        description="This is the user ID. It's between 1 and 100",
        example=1
        )
):
    return {"user_id": user_id}

Overwriting FastAPI/hello_world.py


Pero ¿qué pasa si hacemos un `POST` para añadir un nuevo usuario con su contraseña? Levantamos la API

In [3]:
!uvicorn FastAPI.hello_world:app --reload

INFO:     Will watch for changes in these directories: ['/home/wallabot/Documentos/web/portafolio/posts']
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [17515] using StatReload
INFO:     Started server process [17517]
INFO:     Waiting for application startup.
INFO:     Application startup complete.


Hacemos el `POST`

In [6]:
!curl -X 'POST' \
  'http://127.0.0.1:8000/user/new' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{"user": {"name": "Mariano", "age": 30, "email": "mariano@email.com", "password": "12345678"}, "location": {"city": "Sevilla", "state": "Andalucía", "country": "España"} }'

{"name":"Mariano","age":30,"email":"mariano@email.com","password":"12345678","city":"Sevilla","state":"Andalucía","country":"España"}

Si nos fijamos nos está devolviendo el password del usuario, pero esto es muy pelogroso, ya que su password quedaría expuesto.

Para evitar esto vamos a crear una nueva clase `UserBasic` que va a ser igual que lo que era `User` pero sin password. Y vamos a crear la clase `User` que va a heredar de `UserBasic` y que va a añadir `password`. Definiremos en el decorador del método `POST` que la clase `UserBasic` será lo que devolvamos en la operación `POST`, así no se devolverá el `password`

In [9]:
%%writefile FastAPI/hello_world.py

from typing import Optional
from pydantic import BaseModel
from pydantic import Field
from pydantic import validator
from enum import Enum
from fastapi import FastAPI
from fastapi import Body
from fastapi import Query
from fastapi import Path

app = FastAPI()

class ComunidadAutonoma(str, Enum):
    Andalucía = "Andalucía"
    Aragón = "Aragón"
    Asturias = "Asturias"
    Baleares = "Baleares"
    Canarias = "Canarias"
    Cantabria = "Cantabria"
    CastillaLaMancha = "Castilla-La Mancha"
    CastillaLeon = "Castilla y León"
    Cataluña = "Cataluña"
    Ceuta = "Ceuta"
    Extremadura = "Extremadura"
    Galicia = "Galicia"
    Madrid = "Madrid"
    Melilla = "Melilla"
    Murcia = "Murcia"
    Navarra = "Navarra"
    PaisVasco = "País Vasco"
    LaRioja = "La Rioja"
    Valencia = "Valencia"

class UserBasic(BaseModel):
    name: str = Field(..., min_length=1, max_length=50) # Required
    age: int = Field(..., gt=0, le=100) # Required
    email: Optional[str] = Field(None) # Optional

    class Config:
        schema_extra = {
            "example": {
                "name": "Mariano",
                "age": 30,
                "email": "mariano@email.com"
            },
        }

class User(UserBasic):
    password: str = Field(..., min_length=8, max_length=50) # Required

    class Config:
        schema_extra = {
            "example": {
                "name": "Mariano",
                "age": 30,
                "email": "mariano@email.com",
                "password": "12345678"
            },
        }

class Location(BaseModel):
    city: str = Field(..., min_length=1, max_length=50) # Required
    state: ComunidadAutonoma = Field(..., min_length=1, max_length=50) # Required
    country: str = Field(..., min_length=1, max_length=50) # Required

    @validator("state", pre=True)
    def validate_state(cls, state):
        if state not in ComunidadAutonoma._member_names_:
            raise ValueError(f"Invalid state: {state}")
        return state
    
    class Config:
        schema_extra = {
            "example": {
                "city": "Sevilla",
                "state": "Andalucía",
                "country": "España"
            },
        }

@app.get("/")
def home():
    return {"Hello": "World"}

# Request and Response body
@app.post("/user/new", response_model=UserBasic)
def new_user(
    user: User = Body(...), 
    location: Location = Body(...)
):
    resutl = user.dict()
    resutl.update(location.dict())
    return resutl

# Validaciones de los query parameters
@app.get("/user/detail")
def user_show_detail(
    name: Optional[str] = Query(
        None, 
        min_length=1, 
        max_length=50,
        title="User name",
        description="This is the user name. It's between 1 and 50 characters",
        example="Mariano"
        ),
    age: Optional[int] = Query(
        None, 
        gt=0, 
        le=100,
        title="User age",
        description="This is the user age. It's between 1 and 100",
        example=30
        )
):
    return {"name": name, "age": age}

# Validaciones de los path parameters
@app.get("/user/detail/{user_id}")
def user_show_detail(
    user_id: int = Path(
        ..., 
        gt=0, 
        le=100,
        title="User ID",
        description="This is the user ID. It's between 1 and 100",
        example=1
        )
):
    return {"user_id": user_id}

Overwriting FastAPI/hello_world.py


Levantamos la API

In [ ]:
!uvicorn FastAPI.hello_world:app --reload

INFO:     Will watch for changes in these directories: ['/home/wallabot/Documentos/web/portafolio/posts']
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [17515] using StatReload
INFO:     Started server process [17517]
INFO:     Waiting for application startup.
INFO:     Application startup complete.


Y hacemos el `POST`

In [11]:
!curl -X 'POST' \
  'http://127.0.0.1:8000/user/new' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{"user": {"name": "Mariano", "age": 30, "email": "mariano@email.com", "password": "12345678"}, "location": {"city": "Sevilla", "state": "Andalucía", "country": "España"} }'

{"name":"Mariano","age":30,"email":"mariano@email.com"}

Como vemos ahora no nos devuelve ni el `password` ni la clase `Location`

## Status code

Cuando hacemos un `GET`, un `POST`, etc. Recibimos un código, por ejemplo, si intentamos acceder a una página que no existe recibimos el error 404. A este código se le llama `status code` y los tipos son
 * Respuestas informativas (100-199)
 * Respuestas Satisfactorias (200-299)
 * Redirecciones (300-399)
 * Errores de los clientes (400-499)
 * Errores de los servidores (500-599)

Así que podemos devolver uno de esos valores en nuestra API, para ello importamos el módulo `status` de la librería `fastapi`

In [13]:
%%writefile FastAPI/hello_world.py

from typing import Optional
from pydantic import BaseModel
from pydantic import Field
from pydantic import validator
from enum import Enum
from fastapi import FastAPI
from fastapi import Body
from fastapi import Query
from fastapi import Path
from fastapi import status

app = FastAPI()

class ComunidadAutonoma(str, Enum):
    Andalucía = "Andalucía"
    Aragón = "Aragón"
    Asturias = "Asturias"
    Baleares = "Baleares"
    Canarias = "Canarias"
    Cantabria = "Cantabria"
    CastillaLaMancha = "Castilla-La Mancha"
    CastillaLeon = "Castilla y León"
    Cataluña = "Cataluña"
    Ceuta = "Ceuta"
    Extremadura = "Extremadura"
    Galicia = "Galicia"
    Madrid = "Madrid"
    Melilla = "Melilla"
    Murcia = "Murcia"
    Navarra = "Navarra"
    PaisVasco = "País Vasco"
    LaRioja = "La Rioja"
    Valencia = "Valencia"

class UserBasic(BaseModel):
    name: str = Field(..., min_length=1, max_length=50) # Required
    age: int = Field(..., gt=0, le=100) # Required
    email: Optional[str] = Field(None) # Optional

    class Config:
        schema_extra = {
            "example": {
                "name": "Mariano",
                "age": 30,
                "email": "mariano@email.com"
            },
        }

class User(UserBasic):
    password: str = Field(..., min_length=8, max_length=50) # Required

    class Config:
        schema_extra = {
            "example": {
                "name": "Mariano",
                "age": 30,
                "email": "mariano@email.com",
                "password": "12345678"
            },
        }

class Location(BaseModel):
    city: str = Field(..., min_length=1, max_length=50) # Required
    state: ComunidadAutonoma = Field(..., min_length=1, max_length=50) # Required
    country: str = Field(..., min_length=1, max_length=50) # Required

    @validator("state", pre=True)
    def validate_state(cls, state):
        if state not in ComunidadAutonoma._member_names_:
            raise ValueError(f"Invalid state: {state}")
        return state
    
    class Config:
        schema_extra = {
            "example": {
                "city": "Sevilla",
                "state": "Andalucía",
                "country": "España"
            },
        }

@app.get("/", status_code=status.HTTP_200_OK)
def home():
    return {"Hello": "World"}

# Request and Response body
@app.post("/user/new", response_model=UserBasic, status_code=status.HTTP_201_CREATED)
def new_user(
    user: User = Body(...), 
    location: Location = Body(...)
):
    resutl = user.dict()
    resutl.update(location.dict())
    return resutl

# Validaciones de los query parameters
@app.get("/user/detail", status_code=status.HTTP_200_OK)
def user_show_detail(
    name: Optional[str] = Query(
        None, 
        min_length=1, 
        max_length=50,
        title="User name",
        description="This is the user name. It's between 1 and 50 characters",
        example="Mariano"
        ),
    age: Optional[int] = Query(
        None, 
        gt=0, 
        le=100,
        title="User age",
        description="This is the user age. It's between 1 and 100",
        example=30
        )
):
    return {"name": name, "age": age}

# Validaciones de los path parameters
@app.get("/user/detail/{user_id}", status_code=status.HTTP_200_OK)
def user_show_detail(
    user_id: int = Path(
        ..., 
        gt=0, 
        le=100,
        title="User ID",
        description="This is the user ID. It's between 1 and 100",
        example=1
        )
):
    return {"user_id": user_id}

Overwriting FastAPI/hello_world.py


Como vemos en todos los decoradores hemos añadido el código medainte el atributo `status`, vamos a levantar la API

In [ ]:
!uvicorn FastAPI.hello_world:app --reload

INFO:     Will watch for changes in these directories: ['/home/wallabot/Documentos/web/portafolio/posts']
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [17515] using StatReload
INFO:     Started server process [17517]
INFO:     Waiting for application startup.
INFO:     Application startup complete.


Vamos a hacer una operación para ver el código que obtenemos, para eso añadimos el flag `-i` al `curl`

In [2]:
!curl -i -X 'POST' \
  'http://127.0.0.1:8000/user/new' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{"user": {"name": "Mariano", "age": 30, "email": "mariano@email.com", "password": "12345678"}, "location": {"city": "Sevilla", "state": "Andalucía", "country": "España"} }'

HTTP/1.1 201 Created
date: Wed, 10 May 2023 14:08:53 GMT
server: uvicorn
content-length: 55
content-type: application/json

{"name":"Mariano","age":30,"email":"mariano@email.com"}

Como vemos hemos recibido un 201, si vemos la definición de la función

``` python
# Request and Response body
@app.post("/user/new", response_model=UserBasic, status_code=status.HTTP_201_CREATED)
def new_user(
    user: User = Body(...), 
    location: Location = Body(...)
):
    resutl = user.dict()
    resutl.update(location.dict())
    return resutl
```

Es lo que teníamos que haber recibido

## Formularios

Para trabajar con formularios, lo primero que tenemos que hacer es instalar el `python-multipart` con el comando `conda install -y -c conda-forge python-multipart`

In [3]:
%conda install -y -c conda-forge python-multipart

Retrieving notices: ...working... done
Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 23.1.0
  latest version: 23.3.1

Please update conda by running

    $ conda update -n base -c conda-forge conda

Or to minimize the number of packages updated during conda update use

     conda install conda=23.3.1



## Package Plan ##

  environment location: /home/wallabot/miniconda3/envs/fastapi

  added / updated specs:
    - python-multipart
...
Preparing transaction: done
Verifying transaction: done
Executing transaction: done


Vamos a crear un nuevo formulario con una operación de tipo `POST`

In [9]:
%%writefile FastAPI/hello_world.py

from typing import Optional
from pydantic import BaseModel
from pydantic import Field
from pydantic import validator
from enum import Enum
from fastapi import FastAPI
from fastapi import Body
from fastapi import Query
from fastapi import Path
from fastapi import status
from fastapi import Form

app = FastAPI()

class ComunidadAutonoma(str, Enum):
    Andalucía = "Andalucía"
    Aragón = "Aragón"
    Asturias = "Asturias"
    Baleares = "Baleares"
    Canarias = "Canarias"
    Cantabria = "Cantabria"
    CastillaLaMancha = "Castilla-La Mancha"
    CastillaLeon = "Castilla y León"
    Cataluña = "Cataluña"
    Ceuta = "Ceuta"
    Extremadura = "Extremadura"
    Galicia = "Galicia"
    Madrid = "Madrid"
    Melilla = "Melilla"
    Murcia = "Murcia"
    Navarra = "Navarra"
    PaisVasco = "País Vasco"
    LaRioja = "La Rioja"
    Valencia = "Valencia"

class UserBasic(BaseModel):
    name: str = Field(..., min_length=1, max_length=50) # Required
    age: int = Field(..., gt=0, le=100) # Required
    email: Optional[str] = Field(None) # Optional

    class Config:
        schema_extra = {
            "example": {
                "name": "Mariano",
                "age": 30,
                "email": "mariano@email.com"
            },
        }

class User(UserBasic):
    password: str = Field(..., min_length=8, max_length=50) # Required

    class Config:
        schema_extra = {
            "example": {
                "name": "Mariano",
                "age": 30,
                "email": "mariano@email.com",
                "password": "12345678"
            },
        }

class Location(BaseModel):
    city: str = Field(..., min_length=1, max_length=50) # Required
    state: ComunidadAutonoma = Field(..., min_length=1, max_length=50) # Required
    country: str = Field(..., min_length=1, max_length=50) # Required

    @validator("state", pre=True)
    def validate_state(cls, state):
        if state not in ComunidadAutonoma._member_names_:
            raise ValueError(f"Invalid state: {state}")
        return state
    
    class Config:
        schema_extra = {
            "example": {
                "city": "Sevilla",
                "state": "Andalucía",
                "country": "España"
            },
        }
    
class LoginOut(BaseModel):
    username: str = Field(..., min_length=1, max_length=50) # Required
    message: str = Field(min_length=1, max_length=50) # Required

    class Config:
        schema_extra = {
            "example": {
                "username": "Mariano",
                "message": "Login successful"
            },
        }

@app.get("/", status_code=status.HTTP_200_OK)
def home():
    return {"Hello": "World"}

# Request and Response body
@app.post("/user/new", response_model=UserBasic, status_code=status.HTTP_201_CREATED)
def new_user(
    user: User = Body(...), 
    location: Location = Body(...)
):
    resutl = user.dict()
    resutl.update(location.dict())
    return resutl

# Validaciones de los query parameters
@app.get("/user/detail", status_code=status.HTTP_200_OK)
def user_show_detail(
    name: Optional[str] = Query(
        None, 
        min_length=1, 
        max_length=50,
        title="User name",
        description="This is the user name. It's between 1 and 50 characters",
        example="Mariano"
        ),
    age: Optional[int] = Query(
        None, 
        gt=0, 
        le=100,
        title="User age",
        description="This is the user age. It's between 1 and 100",
        example=30
        )
):
    return {"name": name, "age": age}

# Validaciones de los path parameters
@app.get("/user/detail/{user_id}", status_code=status.HTTP_200_OK)
def user_show_detail(
    user_id: int = Path(
        ..., 
        gt=0, 
        le=100,
        title="User ID",
        description="This is the user ID. It's between 1 and 100",
        example=1
        )
):
    return {"user_id": user_id}

# Formulario
@app.post("/user/new/form", status_code=status.HTTP_201_CREATED, response_model=LoginOut)
def new_user_form(username: str=Form(...), password: str=Form(...)):
    return LoginOut(username=username, message="Login successful")

Overwriting FastAPI/hello_world.py


Primero importamos la clase `Form` de `fastapi`

``` python
from fastapi import Form
```

Ahora creamos la clase que se devolverá al hacer un `POST` a la nueva ruta que vamos a crear para hacer un formulario

``` python
class LoginOut(BaseModel):
    username: str = Field(..., min_length=1, max_length=50) # Required
    message: str = Field(min_length=1, max_length=50) # Required

    class Config:
        schema_extra = {
            "example": {
                "username": "Mariano",
                "message": "Login successful"
            },
        }
```

Y por último creamos la función para el `POST` con el formulario

``` python
# Formulario
@app.post("/user/new/form", status_code=status.HTTP_201_CREATED, response_model=LoginOut)
def new_user_form(username: str=Form(...), password: str=Form(...)):
    return LoginOut(username=username, message="Login successful")
```

Levantamos la API

In [ ]:
!uvicorn FastAPI.hello_world:app --reload

INFO:     Will watch for changes in these directories: ['/home/wallabot/Documentos/web/portafolio/posts']
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [17515] using StatReload
INFO:     Started server process [17517]
INFO:     Waiting for application startup.
INFO:     Application startup complete.


Y hacemos el `POST`

In [11]:
!curl -X 'POST' \
  'http://127.0.0.1:8000/user/new/form' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/x-www-form-urlencoded' \
  -d 'username=mariano&password=1234'

{"username":"mariano","message":"Login successful"}

## Cokie and Header parameters

Vamos a realizar un formulario de contacto en el que el usuario, entre otras cosas nos tiene que dat su email. Así que para ello instalamos una librería que nos sirve para comprobar dicho email

In [3]:
%conda install -c conda-forge email_validator

Retrieving notices: ...working... done
Solving environment: done


## Package Plan ##

  environment location: /home/wallabot/miniconda3/envs/fastapi

  added / updated specs:
    - email_validator
...
Preparing transaction: done
Verifying transaction: done
Executing transaction: done

Note: you may need to restart the kernel to use updated packages.


In [8]:
%%writefile FastAPI/hello_world.py

from typing import Optional
from pydantic import BaseModel
from pydantic import Field
from pydantic import validator
from pydantic import EmailStr
from enum import Enum
from fastapi import FastAPI
from fastapi import Body
from fastapi import Query
from fastapi import Path
from fastapi import status
from fastapi import Form
from fastapi import Cookie, Header

app = FastAPI()

class ComunidadAutonoma(str, Enum):
    Andalucía = "Andalucía"
    Aragón = "Aragón"
    Asturias = "Asturias"
    Baleares = "Baleares"
    Canarias = "Canarias"
    Cantabria = "Cantabria"
    CastillaLaMancha = "Castilla-La Mancha"
    CastillaLeon = "Castilla y León"
    Cataluña = "Cataluña"
    Ceuta = "Ceuta"
    Extremadura = "Extremadura"
    Galicia = "Galicia"
    Madrid = "Madrid"
    Melilla = "Melilla"
    Murcia = "Murcia"
    Navarra = "Navarra"
    PaisVasco = "País Vasco"
    LaRioja = "La Rioja"
    Valencia = "Valencia"

class UserBasic(BaseModel):
    name: str = Field(..., min_length=1, max_length=50) # Required
    age: int = Field(..., gt=0, le=100) # Required
    email: Optional[str] = Field(None) # Optional

    class Config:
        schema_extra = {
            "example": {
                "name": "Mariano",
                "age": 30,
                "email": "mariano@email.com"
            },
        }

class User(UserBasic):
    password: str = Field(..., min_length=8, max_length=50) # Required

    class Config:
        schema_extra = {
            "example": {
                "name": "Mariano",
                "age": 30,
                "email": "mariano@email.com",
                "password": "12345678"
            },
        }

class Location(BaseModel):
    city: str = Field(..., min_length=1, max_length=50) # Required
    state: ComunidadAutonoma = Field(..., min_length=1, max_length=50) # Required
    country: str = Field(..., min_length=1, max_length=50) # Required

    @validator("state", pre=True)
    def validate_state(cls, state):
        if state not in ComunidadAutonoma._member_names_:
            raise ValueError(f"Invalid state: {state}")
        return state
    
    class Config:
        schema_extra = {
            "example": {
                "city": "Sevilla",
                "state": "Andalucía",
                "country": "España"
            },
        }
    
class LoginOut(BaseModel):
    username: str = Field(..., min_length=1, max_length=50) # Required
    message: str = Field(min_length=1, max_length=50) # Required

    class Config:
        schema_extra = {
            "example": {
                "username": "Mariano",
                "message": "Login successful"
            },
        }

@app.get("/", status_code=status.HTTP_200_OK)
def home():
    return {"Hello": "World"}

# Request and Response body
@app.post("/user/new", response_model=UserBasic, status_code=status.HTTP_201_CREATED)
def new_user(
    user: User = Body(...), 
    location: Location = Body(...)
):
    resutl = user.dict()
    resutl.update(location.dict())
    return resutl

# Validaciones de los query parameters
@app.get("/user/detail", status_code=status.HTTP_200_OK)
def user_show_detail(
    name: Optional[str] = Query(
        None, 
        min_length=1, 
        max_length=50,
        title="User name",
        description="This is the user name. It's between 1 and 50 characters",
        example="Mariano"
        ),
    age: Optional[int] = Query(
        None, 
        gt=0, 
        le=100,
        title="User age",
        description="This is the user age. It's between 1 and 100",
        example=30
        )
):
    return {"name": name, "age": age}

# Validaciones de los path parameters
@app.get("/user/detail/{user_id}", status_code=status.HTTP_200_OK)
def user_show_detail(
    user_id: int = Path(
        ..., 
        gt=0, 
        le=100,
        title="User ID",
        description="This is the user ID. It's between 1 and 100",
        example=1
        )
):
    return {"user_id": user_id}

# Formulario
@app.post("/user/new/form", status_code=status.HTTP_201_CREATED, response_model=LoginOut)
def new_user_form(username: str=Form(...), password: str=Form(...)):
    return LoginOut(username=username, message="Login successful")

# Cookies y Headers
@app.post("/contact", status_code=status.HTTP_200_OK)
def contact(
    name: str = Form(..., min_length=1, max_length=50),
    lastname: str = Form(min_length=1, max_length=50),
    email: EmailStr = Form(...),
    phone: str = Form(...),
    message: str = Form(..., min_length=1, max_length=500),
    user_agent: Optional[str] = Header(None),
    ads: Optional[str] = Cookie(None)
):
    return {
        "name": name,
        "lastname": lastname,
        "email": email,
        "phone": phone,
        "message": message,
        "user_agent": user_agent,
        "ads": ads
    }


Overwriting FastAPI/hello_world.py


Con la linea

``` python
from pydantic import EmailStr
```

importamos la librería para poder validar un email que nos introduzca un usuario

Con la línea

``` python
from fastapi import Cookie, Header
```

Importamos las librerías para las cookies y el header

Con el bloque

``` python
# Cookies y Headers
@app.post("/contact", status_code=status.HTTP_200_OK)
def contact(
    name: str = Form(..., min_length=1, max_length=50),
    lastname: str = Form(min_length=1, max_length=50),
    email: EmailStr = Form(...),
    phone: str = Form(...),
    message: str = Form(..., min_length=1, max_length=500),
    user_agent: Optional[str] = Header(None),
    ads: Optional[str] = Cookie(None)
):
    return {
        "name": name,
        "lastname": lastname,
        "email": email,
        "phone": phone,
        "message": message,
        "user_agent": user_agent,
        "ads": ads
    }
```

Realizamos el formulario de contacto en el que un usuario tiene que meter su nombre, apellido, email, teléfono, un mensage y un `header` y una `cookie`

Levantamos la API

In [ ]:
!uvicorn FastAPI.hello_world:app --reload

INFO:     Will watch for changes in these directories: ['/home/wallabot/Documentos/web/portafolio/posts']
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [17515] using StatReload
INFO:     Started server process [17517]
INFO:     Waiting for application startup.
INFO:     Application startup complete.


Y la probamos

In [2]:
!curl -X 'POST' \
  'http://127.0.0.1:8000/contact' \
  -H 'accept: application/json' \
  -H 'user-agent: user-agent' \
  -H 'Cookie: ads=ads' \
  -H 'Content-Type: application/x-www-form-urlencoded' \
  -d 'name=Mariano&lastname=Martinez&email=mariano%40email.com&phone=666666666&message=Hola%20caracola'

{"name":"Mariano","lastname":"Martinez","email":"mariano@email.com","phone":"666666666","message":"Hola caracola","user_agent":"user-agent","ads":"ads"}

## Files

Vamos a crear una `POST` operation para subir una imagen

In [9]:
%%writefile FastAPI/hello_world.py

from typing import Optional
from pydantic import BaseModel
from pydantic import Field
from pydantic import validator
from pydantic import EmailStr
from enum import Enum
from fastapi import FastAPI
from fastapi import Body
from fastapi import Query
from fastapi import Path
from fastapi import status
from fastapi import Form
from fastapi import Cookie, Header
from fastapi import UploadFile, File

app = FastAPI()

class ComunidadAutonoma(str, Enum):
    Andalucía = "Andalucía"
    Aragón = "Aragón"
    Asturias = "Asturias"
    Baleares = "Baleares"
    Canarias = "Canarias"
    Cantabria = "Cantabria"
    CastillaLaMancha = "Castilla-La Mancha"
    CastillaLeon = "Castilla y León"
    Cataluña = "Cataluña"
    Ceuta = "Ceuta"
    Extremadura = "Extremadura"
    Galicia = "Galicia"
    Madrid = "Madrid"
    Melilla = "Melilla"
    Murcia = "Murcia"
    Navarra = "Navarra"
    PaisVasco = "País Vasco"
    LaRioja = "La Rioja"
    Valencia = "Valencia"

class UserBasic(BaseModel):
    name: str = Field(..., min_length=1, max_length=50) # Required
    age: int = Field(..., gt=0, le=100) # Required
    email: Optional[str] = Field(None) # Optional

    class Config:
        schema_extra = {
            "example": {
                "name": "Mariano",
                "age": 30,
                "email": "mariano@email.com"
            },
        }

class User(UserBasic):
    password: str = Field(..., min_length=8, max_length=50) # Required

    class Config:
        schema_extra = {
            "example": {
                "name": "Mariano",
                "age": 30,
                "email": "mariano@email.com",
                "password": "12345678"
            },
        }

class Location(BaseModel):
    city: str = Field(..., min_length=1, max_length=50) # Required
    state: ComunidadAutonoma = Field(..., min_length=1, max_length=50) # Required
    country: str = Field(..., min_length=1, max_length=50) # Required

    @validator("state", pre=True)
    def validate_state(cls, state):
        if state not in ComunidadAutonoma._member_names_:
            raise ValueError(f"Invalid state: {state}")
        return state
    
    class Config:
        schema_extra = {
            "example": {
                "city": "Sevilla",
                "state": "Andalucía",
                "country": "España"
            },
        }
    
class LoginOut(BaseModel):
    username: str = Field(..., min_length=1, max_length=50) # Required
    message: str = Field(min_length=1, max_length=50) # Required

    class Config:
        schema_extra = {
            "example": {
                "username": "Mariano",
                "message": "Login successful"
            },
        }

@app.get("/", status_code=status.HTTP_200_OK)
def home():
    return {"Hello": "World"}

# Request and Response body
@app.post("/user/new", response_model=UserBasic, status_code=status.HTTP_201_CREATED)
def new_user(
    user: User = Body(...), 
    location: Location = Body(...)
):
    resutl = user.dict()
    resutl.update(location.dict())
    return resutl

# Validaciones de los query parameters
@app.get("/user/detail", status_code=status.HTTP_200_OK)
def user_show_detail(
    name: Optional[str] = Query(
        None, 
        min_length=1, 
        max_length=50,
        title="User name",
        description="This is the user name. It's between 1 and 50 characters",
        example="Mariano"
        ),
    age: Optional[int] = Query(
        None, 
        gt=0, 
        le=100,
        title="User age",
        description="This is the user age. It's between 1 and 100",
        example=30
        )
):
    return {"name": name, "age": age}

# Validaciones de los path parameters
@app.get("/user/detail/{user_id}", status_code=status.HTTP_200_OK)
def user_show_detail(
    user_id: int = Path(
        ..., 
        gt=0, 
        le=100,
        title="User ID",
        description="This is the user ID. It's between 1 and 100",
        example=1
        )
):
    return {"user_id": user_id}

# Formulario
@app.post("/user/new/form", status_code=status.HTTP_201_CREATED, response_model=LoginOut)
def new_user_form(username: str=Form(...), password: str=Form(...)):
    return LoginOut(username=username, message="Login successful")

# Cookies y Headers
@app.post("/contact", status_code=status.HTTP_200_OK)
def contact(
    name: str = Form(..., min_length=1, max_length=50),
    lastname: str = Form(min_length=1, max_length=50),
    email: EmailStr = Form(...),
    phone: str = Form(...),
    message: str = Form(..., min_length=1, max_length=500),
    user_agent: Optional[str] = Header(None),
    ads: Optional[str] = Cookie(None)
):
    return {
        "name": name,
        "lastname": lastname,
        "email": email,
        "phone": phone,
        "message": message,
        "user_agent": user_agent,
        "ads": ads
    }

# Files
@app.post("/post-image", status_code=status.HTTP_200_OK)
def post_image(image: UploadFile = File(...)):
    return {
        "filename": image.filename,
        "format": image.content_type,
        "size": round(len(image.file.read())/1024, ndigits=2),
    }

Overwriting FastAPI/hello_world.py


Con la límea

``` python
from fastapi import UploadFile, File
```

Importamos las librerías para poder manejar archivos

Con el bloque

``` python
@app.post("/post-image", status_code=status.HTTP_200_OK)
def post_image(image: UploadFile = File(...)):
    return {
        "filename": image.filename
        "format": image.content_type
        "size": len(image.file.read())
    }
```

Realizamos la operación de subida de la imagen

Levantamos la API

In [ ]:
!uvicorn FastAPI.hello_world:app --reload

INFO:     Will watch for changes in these directories: ['/home/wallabot/Documentos/web/portafolio/posts']
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [17515] using StatReload
INFO:     Started server process [17517]
INFO:     Waiting for application startup.
INFO:     Application startup complete.


Y la probamos

In [10]:
!curl -X 'POST' \
  'http://127.0.0.1:8000/post-image' \
  -H 'accept: application/json' \
  -H 'Content-Type: multipart/form-data' \
  -F 'image=@../images/fastapi_logo.webp;type=image/webp'

{"filename":"fastapi_logo.webp","format":"image/webp","size":5.68}

## HTTP Exceptions

Vamos a pedir la información de una usuario, pero si el usuario no existe devolveremos un error

In [1]:
%%writefile FastAPI/hello_world.py

from typing import Optional
from pydantic import BaseModel
from pydantic import Field
from pydantic import validator
from pydantic import EmailStr
from enum import Enum
from fastapi import FastAPI
from fastapi import Body
from fastapi import Query
from fastapi import Path
from fastapi import status
from fastapi import Form
from fastapi import Cookie, Header
from fastapi import UploadFile, File
from fastapi import HTTPException

app = FastAPI()

class ComunidadAutonoma(str, Enum):
    Andalucía = "Andalucía"
    Aragón = "Aragón"
    Asturias = "Asturias"
    Baleares = "Baleares"
    Canarias = "Canarias"
    Cantabria = "Cantabria"
    CastillaLaMancha = "Castilla-La Mancha"
    CastillaLeon = "Castilla y León"
    Cataluña = "Cataluña"
    Ceuta = "Ceuta"
    Extremadura = "Extremadura"
    Galicia = "Galicia"
    Madrid = "Madrid"
    Melilla = "Melilla"
    Murcia = "Murcia"
    Navarra = "Navarra"
    PaisVasco = "País Vasco"
    LaRioja = "La Rioja"
    Valencia = "Valencia"

class UserBasic(BaseModel):
    name: str = Field(..., min_length=1, max_length=50) # Required
    age: int = Field(..., gt=0, le=100) # Required
    email: Optional[str] = Field(None) # Optional

    class Config:
        schema_extra = {
            "example": {
                "name": "Mariano",
                "age": 30,
                "email": "mariano@email.com"
            },
        }

class User(UserBasic):
    password: str = Field(..., min_length=8, max_length=50) # Required

    class Config:
        schema_extra = {
            "example": {
                "name": "Mariano",
                "age": 30,
                "email": "mariano@email.com",
                "password": "12345678"
            },
        }

class Location(BaseModel):
    city: str = Field(..., min_length=1, max_length=50) # Required
    state: ComunidadAutonoma = Field(..., min_length=1, max_length=50) # Required
    country: str = Field(..., min_length=1, max_length=50) # Required

    @validator("state", pre=True)
    def validate_state(cls, state):
        if state not in ComunidadAutonoma._member_names_:
            raise ValueError(f"Invalid state: {state}")
        return state
    
    class Config:
        schema_extra = {
            "example": {
                "city": "Sevilla",
                "state": "Andalucía",
                "country": "España"
            },
        }
    
class LoginOut(BaseModel):
    username: str = Field(..., min_length=1, max_length=50) # Required
    message: str = Field(min_length=1, max_length=50) # Required

    class Config:
        schema_extra = {
            "example": {
                "username": "Mariano",
                "message": "Login successful"
            },
        }

@app.get("/", status_code=status.HTTP_200_OK)
def home():
    return {"Hello": "World"}

# Request and Response body
@app.post("/user/new", response_model=UserBasic, status_code=status.HTTP_201_CREATED)
def new_user(
    user: User = Body(...), 
    location: Location = Body(...)
):
    resutl = user.dict()
    resutl.update(location.dict())
    return resutl

# Validaciones de los query parameters
@app.get("/user/detail", status_code=status.HTTP_200_OK)
def user_show_detail(
    name: Optional[str] = Query(
        None, 
        min_length=1, 
        max_length=50,
        title="User name",
        description="This is the user name. It's between 1 and 50 characters",
        example="Mariano"
        ),
    age: Optional[int] = Query(
        None, 
        gt=0, 
        le=100,
        title="User age",
        description="This is the user age. It's between 1 and 100",
        example=30
        )
):
    return {"name": name, "age": age}

# Validaciones de los path parameters
@app.get("/user/detail/{user_id}", status_code=status.HTTP_200_OK)
def user_show_detail(
    user_id: int = Path(
        ..., 
        gt=0, 
        le=100,
        title="User ID",
        description="This is the user ID. It's between 1 and 100",
        example=1
        )
):
    return {"user_id": user_id}

# Formulario
@app.post("/user/new/form", status_code=status.HTTP_201_CREATED, response_model=LoginOut)
def new_user_form(username: str=Form(...), password: str=Form(...)):
    return LoginOut(username=username, message="Login successful")

# Cookies y Headers
@app.post("/contact", status_code=status.HTTP_200_OK)
def contact(
    name: str = Form(..., min_length=1, max_length=50),
    lastname: str = Form(min_length=1, max_length=50),
    email: EmailStr = Form(...),
    phone: str = Form(...),
    message: str = Form(..., min_length=1, max_length=500),
    user_agent: Optional[str] = Header(None),
    ads: Optional[str] = Cookie(None)
):
    return {
        "name": name,
        "lastname": lastname,
        "email": email,
        "phone": phone,
        "message": message,
        "user_agent": user_agent,
        "ads": ads
    }

# Files
@app.post("/post-image", status_code=status.HTTP_200_OK)
def post_image(image: UploadFile = File(...)):
    return {
        "filename": image.filename,
        "format": image.content_type,
        "size": round(len(image.file.read())/1024, ndigits=2),
    }

# HTTP Exceptions
@app.get("/user/{user_id}", status_code=status.HTTP_200_OK)
def get_user(user_id: int):
    if user_id == 1:
        return {"name": "Mariano", "age": 30}
    else:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found")


Overwriting FastAPI/hello_world.py


Con la límea

``` python
from fastapi import HTTPException
```

Importamos las librerías para poder manejar las excepciones

Con el bloque

``` python
@app.get("/user/{user_id}", status_code=status.HTTP_200_OK)
def get_user(user_id: int):
    if user_id == 1:
        return {"name": "Mariano", "age": 30}
    else:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found")
```

Realizamos pedimos la información de un usuario a partir de una ID de usuario. Si el ID no es válido se devuelve un 404

Levantamos la API

In [ ]:
!uvicorn FastAPI.hello_world:app --reload

INFO:     Will watch for changes in these directories: ['/home/wallabot/Documentos/web/portafolio/posts']
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [17515] using StatReload
INFO:     Started server process [17517]
INFO:     Waiting for application startup.
INFO:     Application startup complete.


La provamos enviando un ID válido, como 1

In [3]:
!curl -X 'GET' \
  'http://127.0.0.1:8000/user/1' \
  -H 'accept: application/json'

{"name":"Mariano","age":30}

Ahora la provamos enviando un ID inválido, como 2

In [7]:
!curl -X 'GET' \
  'http://127.0.0.1:8000/user/2' \
  -H 'accept: application/json'

{"detail":"User not found"}

Como vemos devuelve un error

## Documentación

### Orden de la documentación

Si ahora mismo entramos a [http://127.0.0.1:8000/docs#/](http://127.0.0.1:8000/docs#/) vemos que la documentación de las funciones aparecen en el mismo orden en el que las hemos definido